tools/codegen/api/autograd.py

```diff
@@ -0,0 +1,76 @@
+from dataclasses import dataclass
+from typing import Optional, Sequence, List, Tuple
+
+from tools.codegen.api.types import *
+from tools.codegen.model import *
+
+# Represents a saved variable invovled in backward calculation.
+# Note that it can be a derived property of an input argument, e.g.:
+# we could save `other.scalar_type()` instead of the entire `other` tensor.
+@dataclass(frozen=True)
+class Variable:
+    # Name of the saved variable.
+    # Suffix is appended if it's derived property, e.g.: `other_scalar_type`
+    name: str
+
+    # The cpp type string.
+    # TODO: change from raw string to model.Type
+    type: str
+
+    # The expression to read the derived property at save time, e.g.:
+    # `other.scalar_type()`.
+    expr: str
+
+# Represents a backward formula that calculates derivatives for one
+# or more tensors.
+@dataclass(frozen=True)
+class Derivative:
+    # The formula string (legit C++ expression).
+    # Note that expressions against input arguments are replaced with the
```

Comment:
nit: it is not super clear if this replacement is already done in this string or if it needs to be done before using it?

Reply:
'are' -> 'have been' and updated the wording below